### ISSUE TITLE
Remove 'static constraint for IO on server::Connection

### ISSUE THREAD

**Issue by iamwwc** [hyperium/hyper#3114]
**Is your feature request related to a problem? Please describe.**
 Does hyper has plan to support &mut T? we are trying to implement http connection pool, reuse connection when request to the same upstream. But serve_connection need T be 'static. After http connection is over, we can not  access underlying io anymore

`serve_connection` only need `I` is `AsyncRead + AsyncWrite`, but Future return by `serve_connection` require `I` to be 'static
https://docs.rs/hyper/latest/hyper/server/conn/struct.Connection.html#impl-Future-for-Connection%3CI%2C%20S%2C%20E%3E

**Describe the solution you'd like**
https://docs.rs/hyper/0.14.23/hyper/server/conn/struct.Http.html#method.serve_connection
serve_connection should take `&mut io` instead ownership with 'static, so does future.

we can remove `'static` constraint 
change
`I: AsyncRead + AsyncWrite + Unpin + 'static` to `I: AsyncRead + AsyncWrite + Unpin`

https://github.com/hyperium/hyper/blob/031425f087219f02a87eea3d01b14e75e35a5209/src/server/conn.rs#L863

 
**Additional context**
nginx keepalive directive
http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive

**Comment by seanmonstar** [comment]
hyper handles keep-alive for you. The `Service` you provide to `serve_connection()` will be called with all requests that the client sends on the IO transport.

**Comment by iamwwc** [comment]
I'm not asking for connection reuse on one connection. We want to save underlying tcp connection for later use.
If we have a request /foo go through connection A, after /foo response received, we can save A. When a new request open, A can be reuse to send request go to /bar

**Comment by iamwwc** [comment]
As a proxy server, we keep many opened connection connected to upstream. When client want to take with upstream, we check does there are opened  tcp connection to special upstream? If it has, use it.

**Comment by seanmonstar** [comment]
You're asking about `hyper::server`, but then what you describe is most often client side, keeping a pool of connections available to send messages.

Anyways, many applications have used hyper as I described to do exactly what you're saying. So I worry that this might be an [XY problem](https://en.wikipedia.org/wiki/XY_problem). Consider hyper's [gateway example](https://github.com/hyperium/hyper/blob/0.14.x/examples/gateway.rs).

**Comment by iamwwc** [comment]
Yes, it should be client side. How I can manually maintain connection instead hyper default client pool

**Comment by iamwwc** [comment]
and for server side, can we remove 'static constraint on I? I want to save connection for later use after one request/response over. I don't want transfer ownership into hyper serve_connection

**Comment by seanmonstar** [comment]
For the client, you can create a `(SendRequest, Connection)` pair, spawn the `Connection` to handle keep-alive, and then just reuse the `SendRequest`, perhaps putting it into a map or list, similar to how hyper's built-in pool does it.

**Comment by iamwwc** [comment]
> and for server side, can we remove 'static constraint on I? I want to save connection for later use after one request/response over. I don't want transfer ownership into hyper serve_connection

Ok, thank you, what about server side ;)

**Comment by seanmonstar** [comment]
I can't recall what requires `'static`. But I would recommend trying it the way I suggested first, before trying to change hyper.

**Comment by iamwwc** [comment]
For server side, Your previous suggestion just don't works. If there are 'static, we can't access io anymore after call serve_connection. As a low level http library, hyper should give app more control which built on top of it.

**Comment by iamwwc** [comment]
We have to use forked hyper like https://blog.cloudflare.com/how-we-built-pingora-the-proxy-that-connects-cloudflare-to-the-internet/ does

**Comment by iamwwc** [comment]
our underlying protocol can be tcp, udp, quic, our company's private protocol even though unixsock which implement asyncread asyncwrite, it was not always spawn from hyper http server. Also, one connection is not always talk with http, in real world case, sometime another protocol can get involved after http over

**Comment by seanmonstar** [comment]
Ok, so I just took a moment and removed that bound on `Future for Connection`, and it seems to still compile just fine. If you want to confirm that is enough for you, I'd welcome and merge a PR doing so.

**Comment by iamwwc** [comment]
I tried in a test case.

For server side, serve_connection() won't finished until tcp closed. so whether `I` is `'static` or not doesn't matter.

serve_connection return pair, which contains io, so `'static` can still be in there.

For client side, `handshake(stream).await.expect("handshake");` take stream ownership, `conn` won't finish until tcp got closed.

I don't know how to get both server/client TcpStream after full http request is over. Does hyper has IO irrelevant mode which only take care http relative logic, after request/resposne over, hyper can return underlying stream(or hyper just borrow &mut T)

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e445a4873a2c554fff639e0cfc32c1ce
